homeassistant/helpers/device_registry.py

```diff
@@ -0,0 +1,74 @@
+"""Provide a way to connect entities belonging to one device.
+"""
+
+import logging
+
+import attr
+
+from homeassistant.core import callback
+from homeassistant.loader import bind_hass
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_REGISTRY = 'device_registry'
+
+
+@attr.s(slots=True, frozen=True)
+class DeviceEntry:
+    """Device Registry Entry."""
+
+    unique_id = attr.ib(type=str)
+    serial = attr.ib(type=str)
+    manufacturer = attr.ib(type=str)
+    model = attr.ib(type=str)
+    connection = attr.ib(type=tuple)
+    sw_version = attr.ib(type=str, default=None)
+    config_entry_id = attr.ib(type=set, default=attr.Factory(set))
+
+
+class DeviceRegistry:
+    """Class to hold a registry of devices."""
+
+    def __init__(self, hass):
+        """Initialize the device registry."""
+        self.hass = hass
+        self.devices = {}
+
+    @callback
+    def async_get_device(self, serial: str, connection: tuple):
+        """Check if device is registered."""
+        for device in self.devices.values():
+            if (device.serial == serial or
```

Comment:
Now we're treating serial as unique ID?

Reply:
Following the discussion up top originally; first it was unique id, it was then moved to serial. Regardless, gonna redo it all to your suggested identifier tuple